src/stan/model/indexing/lvalue_varmat.hpp

```diff
@@ -0,0 +1,396 @@
+                   const U& y, const char* name = "ANON", int depth = 0) {
+  stan::math::check_range("var_vector[uni] assign", name, x.size(),
+                          idxs.head_.n_);
+  const auto coeff_idx = idxs.head_.n_ - 1;
+  double prev_val = x.val().coeffRef(coeff_idx);
+  x.vi_->val_.coeffRef(coeff_idx) = y.val();
+  stan::math::reverse_pass_callback([x, y, coeff_idx, prev_val]() mutable {
+    x.vi_->val_.coeffRef(coeff_idx) = prev_val;
+    prev_val = x.adj().coeffRef(coeff_idx);
+    x.adj().coeffRef(coeff_idx) = 0.0;
+    y.adj() += prev_val;
+  });
+}
+
+/**
+ * Assign to a non-contiguous subset of elements in a vector.
+ *
+ * Types:  vector[multi] <- vector
+ *
+ * @tparam Vec1 `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @tparam Vec2 `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @param[in] x Vector to be assigned.
+ * @param[in] idxs Index holding an `std::vector` of cells to assign to.
+ * @param[in] y Value vector.
+ * @param[in] name Name of variable (default "ANON").
+ * @param[in] depth Indexing depth (default 0).
+ * @throw std::out_of_range If any of the indices are out of bounds.
+ * @throw std::invalid_argument If the value size isn't the same as
+ * the indexed size.
+ */
+template <typename Vec1, typename Vec2,
+          require_all_var_vector_t<Vec1, Vec2>* = nullptr>
+inline void assign(Vec1&& x,
+                   const cons_index_list<index_multi, nil_index_list>& idxs,
+                   const Vec2& y, const char* name = "ANON", int depth = 0) {
+  stan::math::check_size_match("vector[multi] assign", "left hand side",
+                               idxs.head_.ns_.size(), name, y.size());
+  const auto x_size = x.size();
+  const auto assign_size = idxs.head_.ns_.size();
+  arena_t<std::vector<int>> x_idx;
+  arena_t<std::vector<int>> y_idx;
+  // Need to remove duplicates for cases like {2, 3, 2, 2}
+  for (int i = assign_size - 1; i >= 0; --i) {
+    if (!internal::check_duplicate(x_idx, idxs.head_.ns_[i] - 1)) {
+      stan::math::check_range("vector[multi] assign", name, x_size,
+                              idxs.head_.ns_[i]);
+      y_idx.push_back(i);
+      x_idx.push_back(idxs.head_.ns_[i] - 1);
+    }
+  }
+  arena_t<Eigen::Matrix<double, -1, 1>> prev_vals(x_idx.size());
+  for (Eigen::Index i = 0; i < x_idx.size(); ++i) {
+    prev_vals.coeffRef(i) = x.vi_->val_.coeffRef(x_idx[i]);
+  }
+  for (Eigen::Index i = 0; i < x_idx.size(); ++i) {
+    x.vi_->val_.coeffRef(x_idx[i]) = y.vi_->val_.coeff(y_idx[i]);
+  }
+  stan::math::reverse_pass_callback([x, y, x_idx, y_idx, prev_vals]() mutable {
+    for (Eigen::Index i = 0; i < x_idx.size(); ++i) {
+      x.vi_->val_.coeffRef(x_idx[i]) = prev_vals.coeffRef(i);
+      prev_vals.coeffRef(i) = x.adj().coeffRef(x_idx[i]);
+      x.adj().coeffRef(x_idx[i]) = 0.0;
+      y.adj().coeffRef(y_idx[i]) += prev_vals.coeffRef(i);
```

Comment:
Only for the `uni_index` thought right? If we have something like 
```cpp
assign(x, index_multi({1, 2, 3}, x)
```

We would run into the aliasing issue wouldn't we?

Reply:
Oh right, aliasing. Which your code also does not handle. To handle it, you need to split this into two separate loops - one reading from adjoints and another one writing. And add tests for aliasing too.

Reply:
I'm getting confused. For something like

``cpp
  auto x = generate_linear_var_vector<Vec>(5, 1);
  vector<int> ns{1, 4, 1};
  assign(x, index_list(index_multi(ns)), x.segment(0, 3));
  stan::math::sum(x).grad();
  std::cout << "\nx adj post-grad: \n" << x.adj() << "\n";
```

Should the adjoint for `x` be 

```
x adj post-grad: 
0
2
2
0
1
```

if we did a `matrix<var>` here we would be assigning cells `{1, 2}` to `{0, 3}`. So for that sum those vari pointers are used twice each (for `{0 -> 1}` and `{2 -> 3}`). So to go in reverse here for `var<mat>` the cells `{1, 2}` should have a adjoints of 2 since they were used twice and cells `{0, 3}` would have adjoints of zero. 

Is that right? If so then the current form does handle the aliasing. If not then some help here would be appreciated haha

Reply:
In this example there is no aliasing - no value is both read and written in the assignment.

You can instead try this:
```
  auto x = generate_linear_var_vector<Vec>(5, 1);
  vector<int> ns{2, 3, 4};
  std::cout << "\nx val pre-assign: \n" << x.val() << "\n";
  assign(x, index_list(index_multi(ns)), x.segment(0, 3));
  std::cout << "\nx val post-assign: \n" << x.val() << "\n";
  stan::math::sum(x).grad();
  std::cout << "\nx adj post-grad: \n" << x.adj() << "\n";
```

Reply:
Sorry, I posted the code without trying to run it. The example abova also does not show the issue. This on, however, does:
```
   auto x = generate_linear_var_vector<Eigen::VectorXd>(5, 1);
   vector<int> ns{1, 2, 3};
   std::cout << "\nx val pre-assign: \n" << x.val() << "\n";
   assign(x, index_list(index_multi(ns)), x.segment(1, 3));
   std::cout << "\nx val post-assign: \n" << x.val() << "\n";
   stan::math::sum(x).grad();
   std::cout << "\nx adj post-grad: \n" << x.adj() << "\n";
```

Reply:
Ooof yes thanks! I updated all the multi indexing assignment functions to use separate loops for reading and writing on the forward pass and a separate loop for writing to the assigned from matrix on the reverse pass. I think this is the right scheme. 

Also I sort of gave up on being clever for when multi-index is on the right hand side of a double index assign and just evaluated y before looping over it. I started writing out the code to handle the aliasing manually like the rest of the `assign()` overloads but tbh it felt like the answer was more complicated/expensive relative to just evaluating y once before doing the column assignments.